src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.cs

```diff
@@ -2,19 +2,24 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using System.Collections.Generic;
 using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
 using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Collections;
 using Microsoft.CodeAnalysis.Options;
 using Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion;
 using Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion.Data;
 using Roslyn.Utilities;
+
+using RoslynCompletionItem = Microsoft.CodeAnalysis.Completion.CompletionItem;
 using VSCompletionItem = Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion.Data.CompletionItem;
 
 namespace Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.AsyncCompletion
 {
-    internal sealed partial class ItemManager : IAsyncCompletionItemManager
+    internal sealed partial class ItemManager : IAsyncCompletionItemManager2
```

Comment:
can you link me to where thsi API is defined?

Reply:
https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_git/VS-Platform?path=/src/Editor/Language/Def/Language/AsyncCompletion/IAsyncCompletionItemManager2.cs&_a=contents&version=GBmain

Reply:
so, tbh... this api feels super strange to me:

```c#
    public interface IAsyncCompletionItemManager2 : IAsyncCompletionItemManager
    {
        /// <summary>
        /// This method is first called before completion is about to appear.
        /// The result of this method will be used in subsequent invocations of <see cref="IAsyncCompletionItemManager.UpdateCompletionListAsync"/>
        /// <paramref name="session"/> tracks user user's input tracked with <see cref="IAsyncCompletionSession.ApplicableToSpan"/>.
        /// <paramref name="data"/> provides applicable <see cref="AsyncCompletionSessionDataSnapshot.Snapshot"/> and 
        /// </summary>
        /// <param name="session">The active <see cref="IAsyncCompletionSession"/>. See <see cref="IAsyncCompletionSession.TextView"/></param>
        /// <param name="data">Contains properties applicable at the time this method is invoked.</param>
        /// <param name="token">Cancellation token that may interrupt this operation</param>
        /// <returns>Sorted <see cref="CompletionList{CompletionItem}"/> that will be subsequently passed to <see cref="IAsyncCompletionItemManager.UpdateCompletionListAsync"/></returns>
        Task<CompletionList<CompletionItem>> SortCompletionItemListAsync(
            IAsyncCompletionSession session,
            AsyncCompletionSessionInitialDataSnapshot data,
            CancellationToken token);
    }
```

It feels (imo) more natural if the lang server (us) just exposed an appropriate comparer for items.  Then the platform owns pushign the data back and forth from the snapshot object to the final CompletionList they create.

Reply:
WDYT?

Reply:
Yup, and Editor can handle the allocations within its own boundary. I will bring this up with them

Reply:
@AmadeusW

Reply:
@genlu are we waiting on this first?  what's the plan here?

Reply:
No, we are not waiting on this, since this is following the same pattern of an existing API `IAsyncCompletionItemManager.SortCompletionItemListAsync`

Reply:
Yeah, let's not block on this. It's pretty fundamental part of existing public completion API. The only difference of IAsyncCompletionItemManager2 over IAsyncCompletionItemManager is an ability to return pooled CompletionList\<CompletionItem>. It simplifies switching off ImmutableArray without changing much of existing functionality.
@AmadeusW, would be reasonable to allow passing (or injecting through MEF) a comparer instead of implementing this method?

Reply:
Given that `IAsyncCompletionItemManager2` is new, I think we can simply replace the method with a property `IComparer<CompletionItem> CompletiomItemComparer`.

Reply:
```cs
    public interface IAsyncCompletionItemManager2 : IAsyncCompletionItemManager
    {
        IComparer<CompletionItem> CompletiomItemComparer { get; }

        [Obsolete]
        Task<CompletionList<CompletionItem>> SortCompletionItemListAsync(
            IAsyncCompletionSession session,
            AsyncCompletionSessionInitialDataSnapshot data,
            CancellationToken token);
    }
```

Reply:
looks reasonable to me, but I'd need @AmadeusW to make a call here.